rfcs/0003-timestamps-and-ttl.md

```diff
@@ -0,0 +1,363 @@
+necessary (since flatbuffers does not support single-bit values, a byte is the smallest amount of
+data we can store here):
+
+```
+                |-- meta --|
+|  u16    | var |   byte   |    u32    |  var  |
+|---------|-----|----------|-----------|-------|
+| key_len | key | row_type | value_len | value |
+|---------|-----|----------|-----------|-------|
+```
+
+In cases where both `Timestamp` and `TimeToLive` are enabled, the row will look like:
+
+```
+                |-------- meta --------------|
+|  u16    | var |    byte  | i64 |    i64    |    u32    |  var  |
+|---------|-----|----------------------------|-----------|-------|
+| key_len | key | row_type | ts  | expire_ts | value_len | value |
+|---------|-----|----------|-----|-----------|-----------|-------|
+```
+
+Both `ts` and `expire_ts` are encoded as `i64` in milliseconds.
+
+The metadata will be returned in the decoded `KeyValueDeletable` which will be modified:
+
+```rust
+pub struct KeyValueDeletable {
+    pub key: Bytes,
+    pub value: ValueDeletable,
+    pub meta: KeyValueMeta,
+}
+
+pub struct KeyValueMeta {
+    pub ts: Option<i64>,
+    pub expire_ts: Option<i64>,
+}
+```
+
+### Compaction
+
+#### TTL Compaction Filter
+
+We will introduce the concept of a compaction filter, which serves as a mechanism for pluggable
+participation in the compaction process. Specifically, the filter (if configured) will run on each
+key during the compaction process and optionally remove or modify the value.
+
+For the scope of this RFC, we will be introducing a builtin TS/TTL compaction filter. This filter
+will add a tombstone to any value where `db.clock.now() > kv.meta.expire_ts`. Note that
+tombstones will still contain a timestamp see [out of order insertions](#insertion-timestamps). If
+a tombstone is added, the TTL (`expire_ts`) for the row will be cleared, indicating that a future
+run of this filter should not remove the tombstone.
+
+Since this RFC does not cover the public API for compaction filters, this filter will simply run
+on any SST that specifies `meta_features: [Timestamp | TimeToLive]`.
+
+#### Periodic Compaction
+
+If the SST is written with the `TimeToLive` meta feature enabled, the compaction scheduler will
+compare the SST's `SsTableInfo#timestamp` with a newly introduced compaction config for periodic
+compaction time limits:
+
+```rust
+use std::time::Duration;
+
+pub struct CompactorOptions {
+    // ...
+
+    // When there are SSTs in L0 (or Sorted Runs with SSTs in other levels) older than 
+    // `periodic_compaction_interval`, SlateDB will include them in the next scheduled
+    // compaction -- this is a best effort interval and slateDB does not guarantee that
+    // compaction will happen within this interval (implementations of the compaction scheduler
+    // are free to implement jitters or other mechanisms to ensure that SSTs are not 
+    // all compacted simultaneously)
+    periodic_compaction_interval: Duration
```

Comment:
Since every `SST` has a `max_ts`, and there is a `now()`, compaction should be able to figure out if a TTL induced compaction is necessary.
On top of that, is this `periodic_compaction_interval` to control how often are purely TTL based compactions triggered?

Also, wouldn't need to persist some info about when the last such compaction completed?

Reply:
>  On top of that, is this periodic_compaction_interval to control how often are purely TTL based compactions triggered?

Yes, but taken from a slightly different angle. The idea is to make sure that no SST lives longer than the compaction interval without being compacted (think about it that way, instead of enforcing compaction at some interval). This should also answer your other question, which is that `creation_timestamp` is the mechanism we need rather than when the last compaction completed.